src/gps/nmea_parser.cc

```diff
@@ -0,0 +1,137 @@
+    } catch (std::out_of_range &e) {
+        throw NMEAError(e.what());
+    }
+
+    return true;
+}
+
+bool
+NMEAParser::parseCoordinates(const std::string &line, GPSData &data)
+{
+    using meter_t = units::length::meter_t;
+    using degree_t = units::angle::degree_t;
+    using arcminute_t = units::angle::arcminute_t;
+
+    if (!parse(line, "$GNGGA", 9)) {
+        return false;
+    }
+
+    try {
+        parseTimeField(m_Fields[0], data.time); // UTC time
+        data.milliseconds = stoi(m_Fields[0].substr(7, 2));
+
+        data.coordinate.lat = degree_t(std::stod(m_Fields[1].substr(0, 2))) +
+                              arcminute_t(std::stod(m_Fields[1].substr(2, 8)));
+        if (m_Fields[2][0] == 'W') {
+            data.coordinate.lat = -data.coordinate.lat;
+        }
+        data.coordinate.lon = degree_t(std::stod(m_Fields[3].substr(0, 3))) +
+                              arcminute_t(std::stod(m_Fields[3].substr(3, 9)));
+        if (m_Fields[4][0] == 'S') {
+            data.coordinate.lon = -data.coordinate.lon;
+        }
+        data.gpsQuality = static_cast<GPSQuality>(std::stoi(m_Fields[5]));
+        data.numberOfSatellites = std::stoi(m_Fields[6]);
+        data.horizontalDilution = meter_t{ std::stod(m_Fields[7]) };
+        data.coordinate.height = meter_t{ std::stod(m_Fields[8]) };
+    } catch (std::invalid_argument &e) {
+        throw NMEAError(e.what());
+    } catch (std::out_of_range &e) {
+        throw NMEAError(e.what());
+    }
+
+    return true;
+}
+
+bool
+NMEAParser::parse(const std::string &line, const std::string &sentenceId,
+                  size_t numFields)
+{
+    BOB_ASSERT(numFields <= m_Fields.size());
+
+    std::string field;
+    std::istringstream lineStream{ line };
+
+    // The first field indicates what kind of message we have
+    if (!getline(lineStream, field, ',')) {
+        return false;
+    }
+    if (field.empty() || field[0] != '$') {
+        throw NMEAError{ "Sentence doesn't begin with $" };
+    }
+    if (field != sentenceId) {
+        return false;
+    }
+
+    if (line.size() < 3 || line[line.size() - 3] != '*') {
+        throw NMEAError{ "Missing checksum" };
+    }
+
+    // Check that the checksum matches
+    char *p;
+    long reportedChecksum = std::strtoul(&line[line.size() - 2], &p, 16);
+    if (*p != 0) {
+        throw NMEAError{ "Could not parse checksum" };
+    }
+
+    long actualChecksum = 0;
+    for (size_t i = 1; i < line.size() - 3; i++) {
+        actualChecksum ^= line[i];
+    }
+    if (reportedChecksum != actualChecksum) {
+        std::stringstream ss;
+        ss << "Bad checksum (expected: " << std::hex << reportedChecksum
+           << ", got: " << std::hex << actualChecksum << ")";
+        throw NMEAError{ ss.str() };
+    }
+
+    // Extract the remaining fields
+    size_t count = 0;
+    for (; getline(lineStream, field, ',') && count < numFields; count++) {
+        m_Fields[count] = std::move(field);
```

Comment:
I don't think this is a valid ``std::move``. ``field`` is being re-used so you can't/shouldn't steal it's heap-allocated memory!

Reply:
Apparently the ``std::move`` leaves it in a "valid, but undefined state": https://stackoverflow.com/questions/9168823/reusing-a-moved-container/9168917#9168917

So it's ok to reassign to it (as this operation does not assume the ``std::string`` is in a particular state).

Reply:
but, semantically, moving makes no sense so why do it!?

Reply:
I think doing it this way *might* save us a copy (because the allocated memory from ``field`` will always have been transferred when the loop ends). But there won't be much in it performance-wise :laughing: